New group creation

hubudibu opened the pull request:
work in progress, will at some point fix #311

hubudibu committed:
move group creation data to vuex store from local state

hubudibu committed:
new view for step 1: new group name

hubudibu committed:
new view for step 2: group purpose

hubudibu committed:
new view for step 3: mincome

hubudibu committed:
new view for step 4: rules

hubudibu committed:
Revert "move group creation data to vuex store from local state"

This reverts commit 7eb5233f567580b6c6d36d63f4bc3a32669d4290.

hubudibu committed:
remove routes for create group steps

hubudibu committed:
moved creategroup local state under form/ephemeral keys for #297

hubudibu committed:
moved vueAssistant to components folder

hubudibu committed:
testing out VueAssistant

hubudibu committed:
testing out slots

hubudibu committed:
testing out linking values through props and events

hubudibu committed:
createGroup got its own event bus

hubudibu committed:
better value passing to children in createGroup

hubudibu committed:
forget the event bus, say hi to computed setters

hubudibu committed:
moved create group steps to components

hubudibu committed:
create group name component, vueAssistant step version

hubudibu committed:
create group purpose component, vueAssistant step version

hubudibu committed:
create group mincome component, vueAssistant step version

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -235,23 +84,62 @@ export default {
+    }
+  },
+  computed: {
+    views: function () {
```
OK, so, I haven't checked this myself yet, and I could be mistaken, but my gut tells me that:

1. This doesn't need to be a function
2. This doesn't need to be a computed property (have you tried the `data:` field?)
3. For the stuff inside of it, the `get` / `set` stuff of the computed properties is also not necessary

hubudibu replied:
1. yes it does, see the api docs: https://vuejs.org/v2/api/#computed also, it's using `this` inside to use properties defined in data, and this is how that context is bound to the right place

2. and 3.: I did try the data field, which is good for setting an initial value, but what I want to achieve is binding that keeps up-to-date when changed in either the parent or this child component. 
for this I needed to solve how to update `form.groupName` in the parent when the value in the step child component changes.
I tried numerous solutions for that:
- firing an event in the step component and manually propagating through 4 components to get here 🤾‍♀️🤾‍♂️🤾‍♀️🤾‍♂️, and
- creating createGroup's own event bus for a more direct communication between the step and the parent 🚎
that you can see in the previous commits.
both those ways seemed way more cumbersome than this solution, where I directly use the computed value's setter to update data in the parent component.
And you can't use getter/setters for data, only for computed props.

I'm still thinking about creating a little tool that automates this kind of binding and makes it less ugly (I miss React's HOCs), but apart from that, this is the simplest solution I could find. let me know if you still see ways to simplify!

taoeffect replied:
> yes it does, see the api docs: https://vuejs.org/v2/api/#computed also, it's using this inside to use properties defined in data, and this is how that context is bound to the right place

Ah you're right, I think I might have been just used to seeing it written as `views ()` instead of `views: function ()`, and subconsciously forgotten those are the same thing, lol.

OK, this looks good, I will still play with it and see if I can think of any other way, but I'm getting more convinced that you've found the correct approach.

Kudos on the `connect` function, and nice job with making this PR conform to #297. 👍

hubudibu committed:
connect util for binding stuff to vue comps

hubudibu committed:
create group rules step

hubudibu committed:
create group invite step

hubudibu committed:
summary step for group creation

hubudibu committed:
finish button for vueAssistant, finished assistant submits group creation

hubudibu committed:
make vue assistant i18n ready

hubudibu committed:
clean up step impports

hubudibu committed:
connect v2 woorks with whole form object too

hubudibu committed:
redirect to dashboard after successful group creation

hubudibu committed:
no more anonymous components

hubudibu committed:
invite step now checks if user exists

hubudibu committed:
invite ppl at group creation

hubudibu committed:
remove invitee before inviting on invite step

hubudibu committed:
just a bit of cleanup

hubudibu committed:
improved keyboard navigation

hubudibu committed:
privacy settings dummy step

hubudibu commented:
@taoeffect so I got to a "working happy path" stage, check it out if you have the time!
major things remaining: 

- [x] make the summary page editable / add jump back to section to edit links
- [x] clean up validations, make steps validate themselves
- [x] tests!
- [x] styling
- [x] change route based on current step (low prio)

taoeffect commented:
@hubudibu That sounds perfect! 👍

hubudibu committed:
step assistant using router
in this new version vue-router does the routing part, the assistant is only a mixin for stepping through child routes

hubudibu committed:
summary page got edit links

hubudibu committed:
fix 404ing image on nested views

hubudibu committed:
validation in group creation steps
veeValidate had more limitations than benefits, so browser does the validation now

hubudibu committed:
unbreak group creation tests

hubudibu committed:
styles: assistant & first step

hubudibu committed:
styles: basics for simple steps

hubudibu committed:
styles: rules step

hubudibu committed:
styles: summary step

hubudibu committed:
unbreak group creation tests II

hubudibu committed:
create group test chhecks settings too

hubudibu committed:
create group test chhecks invitees too

hubudibu committed:
🎶 It just takes some time

hubudibu commented:
no longer WIP 🎉 
usable, and simpler than a week ago.
there are still a lot to polish, looking forward to hear your thoughts

hubudibu commented:
we have some more open issues that are somehow connected, gathering them here:

- fixes #311 **Update the Group Creation process to match mockups from last hackathon** that was the original issue
- fixes #183 **Make a generic step-by-step setup assistant for the group creation & others** 
_(note this part: "From #93 we concluded that this setup assistant is best done not by creating a custom component, but by re-using the vue-router somehow.")_
- fixes #258 **After group creation + invite, go to group dashboard** 
_we don't have the success message any more, as inviting people is not a separate step but is handled during group creation_
- fixes #310 **Should be able to: skip inviting members + cancel creating a group**
_at least cancelling is now easy by just navigating away, should we make it more explicit with a cancel button somewhere?_

hubudibu committed:
stepAssistant only needs route names

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -3,188 +3,78 @@
+        <transition name="fade" mode="out-in">
+          <router-view :group="form" :validity="validity" @input="(payload) => updateGroupData(payload)">
+          </router-view>
+        </transition>
```
This is brilliant!

Using vue-router's [nested routes](https://router.vuejs.org/en/essentials/nested-routes.html) and changing the VueAssistant to a mixin definitely looks like the most elegant and simplest way to go, great job! 👍 🥇

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -3,188 +3,78 @@
+          <p class="control" v-if="currentStep + 1 === config.steps.length">
+            <button
+              class="button is-success is-large"
+              @click.prevent="submit"
```
This `.prevent` thing is something I asked Tamas about as well.

Is it necessary or will this work without it without causing any problems? My reading of the [event docs](https://vuejs.org/v2/guide/events.html) suggests that `.prevent` is mostly useful in dealing with forms (preventing submit buttons from reloading the page).

If it's not necessary here or on line 16, then remove it.

hubudibu replied:
It was necessary while this was all wrapped in a form, but you are right, I removed the `form` so they are not necessary anymore. removing 👍

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -236,22 +131,82 @@ export default {
+        groupName: false,
+        sharedValues: false,
+        incomeProvided: false
+      },
```
What's going on with this duplication of fields? Remember [DRY](https://code.tutsplus.com/tutorials/3-key-software-principles-you-must-understand--net-25161) if at all possible.

I'm not sure the `validity` key is necessary. Remember we are already using the [vee-validate](https://github.com/baianat/vee-validate) validation plugin. Is it possible to use that with a combination of checking the values in `form` instead?

taoeffect replied:
If you did explore vee-validate and decided it wasn't the right way to go, I'd like to understand what you found out, i.e. the reasons why, and if they're good reasons, then they should be written up as comments right here above line 190.

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -236,22 +131,82 @@ export default {
+          'CreateGroupPrivacy',
+          'CreateGroupInvitees',
+          'CreateGroupSummary'
+        ]
```
Nice.

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -236,22 +131,82 @@ export default {
+      } catch (error) {
+        console.error(error)
+        // TODO: Create More descriptive errors
+        this.errorMsg = L('Failed to Invite Users')
```
This seems like a strange place to put this code...

It would be better, if possible, to keep components as self-contained as possible, and therefore keep the invite code with the invite component, and the group creation code with the group creation component.

hubudibu replied:
with the new workflow, the invitee component is only for gathering the users to be invited once the group is created, and the actual invitation only takes place once all group data is in order and the finish button is clicked, together with group creation. there is also no group to invite people to when we interact with the invitee component. I would definitely want to avoid putting the invitation itself back together with the invitee collection. but I can move this to its own component if you think that is necessary.

taoeffect replied:
OK. I guess this is fine, although we should still at some point in the future think through all the potential error situations here... It feels to me like something might break if something goes wrong. For now I do not see anything obviously critical at least.

taoeffect commented on test/frontend.js:
```diff
@@ -259,24 +259,82 @@ describe('Frontend', function () {
+        .insert('input[name="invitee"]', username + '4')
+        .click('#addButton')
+        .wait('.invitee')
+        .wait(10000)
```
No manual wait times allowed.

hubudibu replied:
🙈  I used it for debugging and forgot to remove. well, this explains the timeouts. removing.

taoeffect commented on frontend/simple/components/CreateGroup/index.js:
```diff
@@ -0,0 +1,7 @@
+export { default as CreateGroupRules } from './CreateGroupRules.vue'
+export { default as CreateGroupPrivacy } from './CreateGroupPrivacy.vue'
+export { default as CreateGroupInvitees } from './CreateGroupInvitees.vue'
+export { default as CreateGroupSummary } from './CreateGroupSummary.vue'
```
Nicely organized.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupRules.vue:
```diff
@@ -0,0 +1,102 @@
+          </div>
+        </div>
+        <p class="title is-5"><i18n>Add Member</i18n></p>
+        <button id="approveToggle" class="button is-link is-6 subtitle" @click="() => toggle('editApprove')"><i18n>Change</i18n></button>
```
I believe `@click="() => toggle('editApprove')"` can be simplified to `@click="toggle('editApprove')"`

EDIT: same goes for everywhere else this is done, i.e. below on line 58 and possibly elsewhere.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupRules.vue:
```diff
@@ -0,0 +1,102 @@
+      </div>
+      <div class="column">
+        <p class="percent" v-if="!form.editRemove">{{ group.memberRemovalPercentage }}%</p>
+        <div class="field" v-if="form.editRemove">
```
Two `v-if`'s next to each other like this, using the same variable, looks like an excellent candidate for moving line 43 below and putting in a `v-else` on it. :)

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -3,188 +3,78 @@
+              </span>
+            </button>
+          </p>
+          <p class="control" v-if="currentStep + 1 === config.steps.length">
```
Another potential candidate for `v-if` => `v-else` (see if there are any others)

taoeffect commented on frontend/simple/index.html:
```diff
@@ -6,7 +6,7 @@
     <meta name="description" content="">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="stylesheet" href="css/main.css">
+    <link rel="stylesheet" href="/simple/css/main.css">
```
Good call.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,115 @@
+    <h1 id="summaryStep" class="title is-1 has-text-centered"><i18n>Review & Finish</i18n></h1>
+
+    <h2 class="title is-6"><i18n>Group name:</i18n></h2>
+    <p class="subtitle is-2" :class="!group.groupName && 'is-danger'">{{ group.groupName || 'No group name set' }}</p>
```
1. This should probably use vee-validate or something else?
2. If it's not in an `<i18n>` tag, it must be in an `L` function.

(Same goes for all other such instances, e.g. below on line 18 and if necessary in other files)

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,115 @@
+    validity: {type: Object}
+  },
+  components: {
+    StatusIcon
```
If it's not being used, it shouldn't be here.

taoeffect commented on frontend/simple/views/CreateGroup.vue:
```diff
@@ -3,188 +3,78 @@
+              <span class="icon">
+                <i class="fa fa-arrow-right"></i>
+              </span>
+            </button>
```
Ideally, using vee-validate, the next button would highlight and select the invalid / unfilled form element in a red outline, and refuse to switch to the next view.

It would also mean you wouldn't have to tell the user `L('We still need some info from you, please go back and fill missing fields')`.

hubudibu replied:
you are right, right now we don't indicate required or erroneous fields anywhere, that should definitely be fixed.

but on the 'refuse to switch to the next view' part: I intentionally went with a more allowing user flow in this case, where a user could look through all the steps at first to familiarize with the process, and then go back to any step to fill it. I see no gain in blocking and stressing the user ahead in the process.

until now, all the questions were visible at the same time, so the user knew what's ahead. also, there, progressing without filling something would have broken group creation. there that was the ideal solution. Now I see room for discussion.

@willstanley what is your opinion on this from the ux point of view?

taoeffect replied:
IMO, the way I've seen other apps and websites do it is: when a required field is not filled in, it is highlighted in red and oftentimes an effect is applied (e.g. focusing the cursor on the field and making the field itself bounce, for example).

Personally, I don't think users will have much stress from that, as they're used to it and expect required fields to be filled in. Also curious what @willstanley thinks.

willstanley replied:
If it’s a required field, I would prevent the user from moving to the next screen until it’s been filled in.

I think going back through the flow to find the field you’ve missed would cause more hassle for the user.

taoeffect commented:
@hubudibu OK, first pass at the review finished! I think it's ready now for you to have a look at the review comments, make some changes based on them, and then once those are pushed let me know and I will actually go and do an entire run through test to verify it works. :)

EDIT: accidentally closed the issue (clicked the wrong button)

hubudibu committed:
remove .prevent from buttons as this is not a form element any more

hubudibu committed:
remove accidental wait

hubudibu committed:
simplify click handlers

hubudibu committed:
v-else

hubudibu committed:
statusicon is no  longer used

hubudibu committed:
v-else fix

taoeffect commented:
![screen shot 2017-11-24 at 6 04 56 pm](https://user-images.githubusercontent.com/138706/33226429-07962de2-d142-11e7-9d31-3341e5af43a7.jpg)

I don't know how much of the design you want to focus on, but one thing that jumps out clearly here is the different size of the back/next buttons.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupMincome.vue:
```diff
@@ -0,0 +1,45 @@
+        />
+      </div>
+    </div>
+    <p><i18n>How much income would you like your group to provide?</i18n></p>
```
The mincome field should also have next to it a popup specifying the currency unit.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupRules.vue:
```diff
@@ -0,0 +1,102 @@
+              :value="group.memberRemovalPercentage"
+              @input="update"
+            />
+          </div>
```
What do you think about this: instead of having the "change" link turn the textfield into an input field, have it reveal a slider below the numbers that the user can click and drag to adjust the number?

taoeffect replied:
Also: I'm reminded by this quote from our post on *[Deprecating Majority Rule](https://groupincome.org/2016/09/deprecating-mays-theorem/):*

> Therefore, it is dangerous to use majority rule whenever supermajority rule should be used instead, e.g. passing laws. That is why we plan on making supermajority the default threshold for removing members in Group Income, since doing so can have a potentially significant impact on a person’s life.

We should make it so that a function is called when the slider is dragged, and when it's dragged to a "bad place", Group Income will be able to make a warning field appear that explains to them that they're choosing a value that is most likely too low, and even linking them to our blog posts, etc.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,110 @@
+    <hr>
+
+    <h2 class="title is-6"><i18n>Group purpose:</i18n></h2>
+    <p class="subtitle is-2" :class="!group.sharedValues && 'is-danger'">{{ group.sharedValues || 'No group purpose set' }}</p>
```
The font here is too big for the group purpose to be displayed (which can be quite long potentially).

Also, don't forget that the error text here needs to be localized (see comment above).

Honestly, I think that, as per the comments elsehwere, we shouldn't even have the option of saying `'No group purpose set'`, it should just always be set by the time this view is displayed, IMO.

hubudibu committed:
install vuelidate for validation

hubudibu committed:
try vuelidate for group name validation

hubudibu committed:
vuelidate all steps for required

hubudibu committed:
disable next button if step is invalid

hubudibu committed:
advanced validation parameters

hubudibu committed:
i18n on summary page

hubudibu commented:
@taoeffect I tried out **[vuelidate](https://monterail.github.io/vuelidate)** for validation, if you agree that this is a good direction then I will go through the codebase and replace all existing validation with this as of #334 

About **validation**, the only thing remaining is an enter guard for routes of steps after an invalid one, let's talk at today's call whether it's necessary.

I addressed most of the comments too, the only things I see **remaining** are
- [x] currency unit selector on minimum income step and
- [x] rule change slider with info notification on 'majority vote' when too low - I love this idea!

For the design questions, I would consider them low priority until we have progress on #327, I assume lot will change here too when that one moves forward. Let me know if you see something that damages usability and should be still done.

Happy New Year! 🥂

hubudibu committed:
currency dropdown

hubudibu committed:
save mincome currency

hubudibu committed:
dynamic currency list

taoeffect commented:
@hubudibu Slack seems to be having some down time today, but I posted a reply yesterday (a little late in the day).

Basically, yeah, I love your idea to use the circle slider around the numbers, that sounds really fun and cool! 😄 

And if you want to chat about anything, you can always send me a DM on twitter, or post to the gitter, which thankfully is still up: https://gitter.im/okTurtles/group-income

hubudibu committed:
install circle slider for rule percentage setting

hubudibu committed:
circle sliders with majority warning

hubudibu committed:
tests updated to skip rule setting

hubudibu committed:
bigger back button

hubudibu commented:
<img width="710" alt="screen shot 2018-01-10 at 2 07 28 pm" src="https://user-images.githubusercontent.com/1003523/34774477-c96e47c8-f60f-11e7-83c0-3eae4d509d63.png">

new slider circles!
I removed the 'change' button solely because it looks cool to have the circles always showing. however, I'm not sure the interaction needed to change the values - just grab the knob - is obvious for every user this way. @taoeffect let me know what you think about this. it would be great to user-test it :) 

now my only problem is that the rule setting step is not tested, and neither is the circle slider package 😱 I'm trying to come up with a solution.

in the meantime, looking forward to your review @taoeffect !

taoeffect commented on frontend/simple/components/GroupsMinIncome.vue:
```diff
@@ -1,7 +1,7 @@
   <section class="has-text-right">
     <p class="min-income-label">Min Income</p>
-    <p class="min-income">${{ group.incomeProvided}}</p>
+    <p class="min-income"><cyy>{{ group.incomeCurrency }}</cyy>{{ group.incomeProvided }}</p>
```
I have a few concerns about this `cyy` component:

1. It doesn't really do much, so it seems unnecessary (plus its name is unclear)
2. Because it doesn't use bindings, it's not useful in situations where the input currency changes
3. As far as I can tell, it would probably be much simpler to just use a computed property that passes `group.incomeCurrency` through the `symbol` function instead of a `<cyy>` component

hubudibu replied:
you are right, they don't make much sense like this. I had some confusion about how the slots are supposed to work, and just copied some of our i18n component's workings here, but now that I checked it again, yes, the `beforeMount` in cyy of course won't run again and update the value when it changes in the slot.
changing to computed properties.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,116 @@
+
+    <h2 class="title is-6"><i18n>Group name:</i18n></h2>
+    <p class="subtitle is-2" v-if="group.groupName">{{ group.groupName }}</p>
+    <p class="subtitle is-2 is-danger" v-else><i18n>No group name set</i18n></p>
```
This `v-if`/`v-else` condition shouldn't be necessary, right? By this point the validations should guarantee that `group.groupName` is set?

hubudibu replied:
well, as I mentioned, all steps having their own URL makes it pretty easy to end up at this step, even with a simple refresh, that's why I kept these. is that something we should worry about?

taoeffect replied:
Hmm, technically you are correct, if the user is using the app in a browser, they could visit this page somehow... without setting the groupName... theoretically, perhaps by looking at the browsing history.

I have two suggestions on how to deal with this:

1. Just make sure that vuelidate doesn't allow the form to be submitted (I'm guessing it already does that? Can you verify?)
2. Have a guard of some sort, perhaps in `mounted ()`, i.e. maybe a mixin that checks the variables that need to be set by this step, and if they aren't set revert to the beginning of the StepAssistant.

**EDIT:** OK, I see that (1) above is the behavior. I guess that is fine. :)

hubudibu replied:
1. yes, that's the current behavior :)

2. here are my two suggestions:

2.1 the full solution would be to check for every step that all the previous steps are valid, and if not, go back to the first invalid step. this seems like a bigger task, but we have everything for this to work - the step-validator relation and the order of steps
2.2 the quicker solution would be to only guard the summary step, and if it's not valid, jump back to the first step of the assistant. this could work if we assumed that the only way to end up here in an invalid state is by skipping all the previous steps. but it's not the case, someone could easily hit refresh at the 2nd step and only miss the first one. from the user's perspective, without a proper error notification solution #175 , I think this is way more confusing and frustrating than seeing on the summary which step is actually missing and having to manually go back to correct it.

so in my opinion, what we have now is better than 2.2, and I'd love to do 2.1, but I assume we have things with higher priority.
thoughts? :)

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,116 @@
+
+    <h2 class="title is-6"><i18n>Group purpose:</i18n></h2>
+    <p class="title is-4" v-if="group.sharedValues">{{ group.sharedValues }}</p>
+    <p class="subtitle is-2 is-danger" v-else><i18n>No group purpose set</i18n></p>
```
Same concern here as above.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,116 @@
+
+    <h2 class="title is-6"><i18n>Minimum income:</i18n></h2>
+    <p class="subtitle is-2" v-if="group.incomeProvided"><cyy>{{ group.incomeCurrency }}</cyy>{{ group.incomeProvided }}</p>
+    <p class="subtitle is-2 is-danger" v-else><i18n>No group income set</i18n></p>
```
Same as above + other comment below about `<cyy>` component.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupSummary.vue:
```diff
@@ -0,0 +1,116 @@
+    <hr>
+
+    <h2 class="title is-6"><i18n>Members to invite:</i18n></h2>
+    <p v-if="!group.invitees.length" class="subtitle is-2 is-warning"><i18n>Noone here yet</i18n></p>
```
Should be `"No one invited"` I think.

taoeffect commented on frontend/simple/components/CreateGroup/CreateGroupRules.vue:
```diff
@@ -0,0 +1,155 @@
+          on a person’s life. Please consider using a supermajority threshold.
+          <a href="https://groupincome.org/2016/09/deprecating-mays-theorem/#when-majority-rule-can-harm">
+            Read more on our blog about the dangers of majority rule.
+          </a>
```
Nice, but remember to put all text, including this, into `<i18n>` tags. ^_^

hubudibu replied:
shoot!
